Remove SystemJS usage from the development viewer and the unit-tests

Snuffleupagus opened the pull request:
*Please refer to the individual commit messages for additional details.*

It's highly recommended to review this *one* commit at a time, perhaps also ignoring whitespace-only changes where appropriate.
The second commit is probably more-or-less impossible to actually review, so we're probably going to have to trust our test-suite and that I wrote proper regular expressions.

This still leaves SystemJS usage in the font-tests, however fixing that will require larger changes and it thus seemed best to defer that to later since this PR is already very large.

Snuffleupagus committed:
Change `src/core/{glyphlist, unicode}.js` to use standard `import`/`export` statements

While the *built* `pdf.worker.js` file still works correctly with these changes, despite these two files being excluded by Babel[1], the development viewer does not because of issues with SystemJS[2] and/or its Babel-plugin (both of which are old).
Furthermore, note also that excluding these two files from Babel-processing isn't *generally* necessary since e.g. the `gulp mozcentral` command works anyway. The explanation is rather that it's actually the source-map generation which fails for these huge sequences when building the `pdf.worker.js` file.

However, not using standard `import`/`export` statements in all files means we also need to use SystemJS when e.e. running the unit-tests. This is very unfortunate, since SystemJS (or its old Babel-version) doesn't support modern ECMAScript features such as e.g. optional chaining and nullish coalescing.

Unfortunately it also seems that https://bugzilla.mozilla.org/show_bug.cgi?id=1247687, which tracks the implementation of worker-modules in Firefox, has stalled since there hasn't been any updates for six months now.

To hopefully address all of the above, this patch is the first in a series that attempts to further reduce our reliance on SystemJS.

---
[1] The only difference being how the dependencies are handled, in the Webpack-bundled file.

[2] Parsing takes way too long and consumes too much memory, thus rendering the development viewer essentially unusable.

Snuffleupagus committed:
Add a `getArrayLookupTableFactory` helper function and use it to re-format `src/core/{glyphlist, unicode}.js`

*Please note:* Once https://bugzilla.mozilla.org/show_bug.cgi?id=1247687 is implemented, and we've removed SystemJS completely, this entire patch can (and even should) be reverted.

This is similar to the existing `getLookupTableFactory` helper function, but is implemented as outlined in issue 6774.
The re-formatting of the tables were done automatically, by using find-and-replace with regular expressions.

For reasons that I don't even pretend to understand, using this particular structure for these *very* long lookup tables allow SystemJS to process the files correctly/quickly and the development viewer thus works as intended.

Snuffleupagus committed:
Only ignore the `src/core/{glyphlist, unicode}.js` files, during building of `pdf.worker.js`, when source-maps are enabled

This produces a slightly smaller built `pdf.worker.js` file, for e.g. the `gulp mozcentral` build-target.

Snuffleupagus committed:
Load all unit-tests with native `import`, rather than SystemJS

Snuffleupagus committed:
Add (some) optional chaining usage in `src/display/api.js`

Since we no longer use SystemJS to load the unit-tests, there's now nothing that prevents us from using optional chaining and nullish coalescing in the `src/display/` directory.

Snuffleupagus committed:
Load the fake worker, in non-`PRODUCTION` mode, with native async `import`

This removes the last SystemJS usage from both the API and the default viewer.

Snuffleupagus commented:
/botio-linux preview

Snuffleupagus commented:
/botio unittest

timvandermeij commented on gulpfile.js:
```diff
@@ -185,6 +185,19 @@ function createWebpackConfig(defines, output) {
+  if (enableSourceMaps) {
+    babelExcludes.push("src[\\\\\\/]core[\\\\\\/](glyphlist|unicode)");
+  }
+  const babelExcludeRegExp = new RegExp("(" + babelExcludes.join("|") + ")");
```
I think we can use a template string here for slightly improved readability.

Snuffleupagus replied:
Sure, why not.

timvandermeij commented on gulpfile.js:
```diff
@@ -185,6 +185,19 @@ function createWebpackConfig(defines, output) {
+  // `src/core/{glyphlist,unicode}.js` (Babel is too slow for those when
+  // source-maps are enabled) should be excluded from processing.
+  const babelExcludes = [
+    "node_modules[\\\\\\/]core-js",
```
I'm going to assume that, like the second commit, the number of backslashes here is correct, although it looks a bit much :-)

Snuffleupagus replied:
For `gulp mozcentral` the resulting RegExp looks as follows, which is consistent with the old format:
```
/(node_modules[\\\/]core-js|node_modules[\\\/]web-streams-polyfill)/
```
For `gulp generic` the resulting RegExp looks as follows, which is also consistent with the old format:
```
/(node_modules[\\\/]core-js|node_modules[\\\/]web-streams-polyfill|src[\\\/]core[\\\/](glyphlist|unicode))/
```

Since we now build the regular expressing using "normal" strings, we need to do more escaping to not lose part of it.

timvandermeij commented on test/unit/jasmine-boot.js:
```diff
@@ -197,26 +199,26 @@ function initializePDFJS(callback) {
-    if (currentWindowOnload) {
-      currentWindowOnload();
+  function extend(destination, source) {
+    for (var property in source) {
```
`var` -> `const` now that my PR just landed (which created a merge conflict that needs to be addressed; sorry for that!)

Snuffleupagus replied:
No worries!

timvandermeij reviewed (approved):
Looks good with these comments addressed. Nice to see yet another part of SystemJS go!

Snuffleupagus commented:
/botio test

Snuffleupagus committed:
Prevent use of optional chaining and nullish coalescing in the `src/shared/` folder

Given that this code is used on the worker-thread, where SystemJS is still used during development, we need to (for now) handle this folder the same way as the `src/core/` one.

timvandermeij commented:
Nice work!
